bson/decoder_test.go

```diff
@@ -235,6 +235,33 @@ func TestDecoderv2(t *testing.T) {
 			t.Fatalf("Decode error mismatch; expected %v, got %v", ErrDecodeToNil, err)
 		}
 	})
+	t.Run("SetDocumentDecodeType", func(t *testing.T) {
+		type someMap map[string]interface{}
+
+		in := make(someMap)
+		in["foo"] = map[string]interface{}{"bar": "baz"}
+		inType := reflect.TypeOf(in).String()
+		inFooType := reflect.TypeOf(in["foo"]).String()
+
+		bytes, err := Marshal(in)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var bsonOut someMap
+		dc := new(bsoncodec.DecodeContext).SetDocumentDecodeType(reflect.TypeOf(map[string]interface{}{}))
+		dec, _ := NewDecoderWithContext(*dc, bsonrw.NewBSONDocumentReader(bytes))
+		if err := dec.Decode(&bsonOut); err != nil {
+			t.Fatal(err)
+		}
+
+		bsonOutType := reflect.TypeOf(bsonOut).String()
+		assert.Equal(t, inType, bsonOutType, "expected %v to equal %v", inType, bsonOutType)
+
+		bsonFooOutType := reflect.TypeOf(bsonOut["foo"]).String()
+		documentType := reflect.TypeOf(map[string]interface{}{}).String()
+		assert.Equal(t, documentType, bsonFooOutType, "expected %v to equal %v", inFooType, bsonFooOutType)
+	})
```

Comment:
We should add some test cases:
* Checks what happens when we `SetDocumentDecodeType` and the `Decode` output parameter is type `interface{}`. Do we expect the `SetDocumentDecodeType` type to be used when decoding the top-level BSON document as well?
* Check what happens when we `SetDocumentDecodeType` to a `string` or something that can't represent a BSON document. Does unmarshalling return an error?

Reply:
- I would expect that if the output type is an empty interface, or a map pointing to an empty interface, that the logic would convert the top-level type.
- Absolutely, will do.

Reply:
Done!